cannot generate COFFEE from <class 'gem.gem.Delta'>

[demo-beam3d_ufl.zip](https://github.com/firedrakeproject/tsfc/files/2895722/demo-beam3d_ufl.zip)
The attached file, when compiled with `ffc -l dolfin -r tsfc demo-beam3d_ufl.ufl` fails because of `AssertionError: cannot generate COFFEE from <class 'gem.gem.Delta'>`, while uflacs digests it without errors.

--- a/tests/test_tsfc_182.py
+++ b/tests/test_tsfc_182.py
@@ -1,0 +1,33 @@
+import pytest
+
+from ufl import (Coefficient, FiniteElement, MixedElement,
+                 TestFunction, VectorElement, dx, inner, tetrahedron)
+
+from tsfc import compile_form
+
+
+@pytest.mark.parametrize('mode', ['vanilla', 'coffee', 'spectral'])
+def test_delta_elimination(mode):
+    # Code sample courtesy of Marco Morandini:
+    # https://github.com/firedrakeproject/tsfc/issues/182
+    scheme = "default"
+    degree = 3
+
+    element_lambda = FiniteElement("Quadrature", tetrahedron, degree,
+                                   quad_scheme=scheme)
+    element_eps_p = VectorElement("Quadrature", tetrahedron, degree,
+                                  dim=6, quad_scheme=scheme)
+
+    element_chi_lambda = MixedElement(element_eps_p, element_lambda)
+
+    chi_lambda = Coefficient(element_chi_lambda)
+    delta_chi_lambda = TestFunction(element_chi_lambda)
+
+    L = inner(delta_chi_lambda, chi_lambda) * dx(degree=degree, scheme=scheme)
+    kernel, = compile_form(L, parameters={'mode': mode})
+
+
+if __name__ == "__main__":
+    import os
+    import sys
+    pytest.main(args=[os.path.abspath(__file__)] + sys.argv[1:])

--- a/tsfc/coffee_mode.py
+++ b/tsfc/coffee_mode.py
@@ -1,6 +1,6 @@
 from functools import partial, reduce
 
-from gem.node import traversal
+from gem.node import traversal, Memoizer
 from gem.gem import Failure, Sum, index_sum
 from gem.optimise import replace_division, unroll_indexsum
 from gem.refactorise import collect_monomials
@@ -75,6 +75,7 @@ def optimise_expressions(expressions, argument_indices):
             return expressions
 
     # Apply argument factorisation unconditionally
-    classifier = partial(spectral.classify, set(argument_indices))
+    classifier = partial(spectral.classify, set(argument_indices),
+                         delta_inside=Memoizer(spectral._delta_inside))
     monomial_sums = collect_monomials(expressions, classifier)
     return [optimise_monomial_sum(ms, argument_indices) for ms in monomial_sums]

--- a/tsfc/spectral.py
+++ b/tsfc/spectral.py
@@ -3,6 +3,7 @@ from functools import partial, reduce
 from itertools import chain, zip_longest
 
 from gem.gem import Delta, Indexed, Sum, index_sum, one
+from gem.node import Memoizer
 from gem.optimise import delta_elimination as _delta_elimination
 from gem.optimise import remove_componenttensors, replace_division, unroll_indexsum
 from gem.refactorise import ATOMIC, COMPOUND, OTHER, MonomialSum, collect_monomials
@@ -44,6 +45,12 @@ def Integrals(expressions, quadrature_multiindex, argument_multiindices, paramet
     return [Integral(e, quadrature_multiindex, argument_indices) for e in expressions]
 
 
+def _delta_inside(node, self):
+    """Does node contain a Delta?"""
+    return any(isinstance(child, Delta) or self(child)
+               for child in node.children)
+
+
 def flatten(var_reps, index_cache):
     quadrature_indices = OrderedDict()
 
@@ -72,6 +79,7 @@ def flatten(var_reps, index_cache):
         variable, expression = pair
         return frozenset(variable.free_indices)
 
+    delta_inside = Memoizer(_delta_inside)
     # Variable ordering after delta cancellation
     narrow_variables = OrderedDict()
     # Assignments are variable -> MonomialSum map
@@ -80,7 +88,7 @@ def flatten(var_reps, index_cache):
     for free_indices, pair_group in groupby(pairs, group_key):
         variables, expressions = zip(*pair_group)
         # Argument factorise expressions
-        classifier = partial(classify, set(free_indices))
+        classifier = partial(classify, set(free_indices), delta_inside=delta_inside)
         monomial_sums = collect_monomials(expressions, classifier)
         # For each monomial, apply delta cancellation and insert
         # result into delta_simplified.
@@ -111,13 +119,13 @@ def flatten(var_reps, index_cache):
 finalise_options = dict(replace_delta=False)
 
 
-def classify(argument_indices, expression):
+def classify(argument_indices, expression, delta_inside):
     """Classifier for argument factorisation"""
     n = len(argument_indices.intersection(expression.free_indices))
     if n == 0:
         return OTHER
     elif n == 1:
-        if isinstance(expression, (Delta, Indexed)):
+        if isinstance(expression, (Delta, Indexed)) and not delta_inside(expression):
             return ATOMIC
         else:
             return COMPOUND

